### PR TITLE
Handle IOV_MAX limit in RandomAccess

### LINKED ISSUE #108383: RandomAccess.Write throws System.IO.IOException: Invalid argument when buffers count is greater than 1024
### Description

Even though this exception sounds very similar to https://github.com/dotnet/runtime/issues/108322#issue-2552756015 , it is actually a very different issue. It occurs in the Unix implementation of the `RandomAccess.WriteGatherAtOffset` method
This is the stack trace
```
System.IO.IOException: Invalid argument
at System.IO.RandomAccess.WriteGatherAtOffset(SafeFileHandle handle, IReadOnlyList`1 buffers, Int64 fileOffset)
```

Looking at the implementation I have noticed that we call `Interop.Sys.PWriteV` passing the `buffersCount` as a parameter

https://github.com/dotnet/runtime/blob/01aa3d96bb2160144f167b1065e081521d133b48/src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Unix.cs#L199-L202

which in turn calls the native implementation
https://github.com/dotnet/runtime/blob/c049b850a75ccf654fb762d89e0132fd999e4cc1/src/native/libs/System.Native/pal_io.c#L2000

According to the documentation of `pwritev` here http://man.he.net/man2/pwritev we **cannot** pass `buffersCount` greater than **1024**

```
NOTES
       POSIX.1  allows  an  implementation  to  place a limit on the number of
       items that can be passed in iov.  An implementation can  advertise  its
       limit  by  defining IOV_MAX in <limits.h> or at run time via the return
       value from sysconf(_SC_IOV_MAX).  On modern Linux systems, the limit is
       1024.  Back in Linux 2.0 days, this limit was 16.
```

### Reproduction Steps

Call `RandomAccess.Write(handle, buffers, fileOffset)` by passing list of buffers where buffers are more than 1024

### Expected behavior

It should save the buffers to file

### Actual behavior

It throws `System.IO.IOException: Invalid argument`

### Regression?

Not tested on .NET6

### Known Workarounds

You can split the buffers into smaller lists (<=1024 elements) and call `RandomAccess.Write` multiple times with the proper offset

### Configuration

Which version of .NET is the code running on?
**.NET7**
What OS and version, and what distro if applicable?
**Ubuntu 22.04.3 LTS under WSL2**
What is the architecture (x64, x86, ARM, ARM64)?
**x64**
Do you know whether it is specific to that configuration?
**It is specific to the Unix implementation of the** `RandomAccess.WriteGatherAtOffset` **method**

### Other information

Haven't tested it, but this should also affect `Read(SafeFileHandle handle, IReadOnlyList<Memory<byte>> buffers, long fileOffset)` which calls `ReadScatterAtOffset(SafeFileHandle handle, IReadOnlyList<Memory<byte>> buffers, long fileOffset)`

### PR DESCRIPTION
**Pull request by adamsitnik**
We were not respecting IOV_MAX in `pwritev` and `preadv` exposed by `RandomAccess`. After writing test for it I've realized that the logic for handling the incomplete writes was also incomplete (please see the changes in managed layer).

fixes #108383 and adds a test for scenario described in https://github.com/dotnet/runtime/issues/108322 , so I can backport https://github.com/dotnet/runtime/pull/108380 to all supported versions.

### CODE REVIEW

File: src/native/libs/System.Native/pal_io.c [hunk 1]
```diff
@@ -1949,6 +1949,22 @@ int32_t SystemNative_PWrite(intptr_t fd, void* buffer, int32_t bufferSize, int64
...
+{
+    int allowedCount = (int)vectorCount;
+
+#if defined(IOV_MAX)
+    if (IOV_MAX < allowedCount)
+    {
+        // We need to respect the limit of items that can be passed in iov.
+        // In case of writes, the managed code is reponsible for handling incomplete writes.
+        // In case of reads, we simply returns the number of bytes read and it's up to the users.
+        allowedCount = IOV_MAX;
```
Comment (MichalPetryka) [inline src/native/libs/System.Native/pal_io.c:1962]: Shouldn't `sysconf(_SC_IOV_MAX)` be preferred here?

Comment (adamsitnik) [reply src/native/libs/System.Native/pal_io.c:1962]: > Shouldn't `sysconf(_SC_IOV_MAX)` be preferred here?

Great question! I could definitely read it once and store in a static field.

@tmds what is your recommendation?

Comment (tmds) [reply src/native/libs/System.Native/pal_io.c:1962]: I think it's fine to use `IOV_MAX` here. This means we're deferring binary compatibility to the underlying C library like we're doing as well when compiling with an older glibc version and relying that to work on newer glibc versions.

Side-notes:
- The limit is 1024 on Linux. If you are hitting this limit, probably you application would benefit from using larger buffers.
- The Linux kernel has some optimizations (stackallocs) for <= 8 iovs (`UIO_FASTIOV`).
- POSIX says `IOV_MAX` should be at least 16 (`_POSIX_UIO_MAXIOV`).

File: src/native/libs/System.Native/pal_io.c [hunk 2]
```diff
@@ -1949,6 +1949,22 @@ int32_t SystemNative_PWrite(intptr_t fd, void* buffer, int32_t bufferSize, int64
     return (int32_t)count;
 }
 
+static int GetAllowedVectorCount(int32_t vectorCount)
```
Comment (am11) [inline src/native/libs/System.Native/pal_io.c:1952]:
```suggestion
#if (HAVE_PREADV || HAVE_PWRITEV) && !defined(TARGET_WASM)
static int GetAllowedVectorCount(int32_t vectorCount)
```

File: src/native/libs/System.Native/pal_io.c [hunk 3]
```diff
@@ -1949,6 +1949,22 @@ int32_t SystemNative_PWrite(intptr_t fd, void* buffer, int32_t bufferSize, int64
...
+    if (IOV_MAX < allowedCount)
+    {
+        // We need to respect the limit of items that can be passed in iov.
+        // In case of writes, the managed code is reponsible for handling incomplete writes.
+        // In case of reads, we simply returns the number of bytes read and it's up to the users.
+        allowedCount = IOV_MAX;
+    }
+#endif
+    return allowedCount;
+}
```
Comment (am11) [inline src/native/libs/System.Native/pal_io.c:1966]:
```suggestion
}
#endif // (HAVE_PREADV || HAVE_PWRITEV) && !defined(TARGET_WASM)
```

File: src/native/libs/System.Native/pal_io.c (hunk 2 above)
Comment (adamsitnik) [reply src/native/libs/System.Native/pal_io.c:1953]: @am11 thanks a lot!

File: src/native/libs/System.Native/pal_io.c (hunk 1 above)
Comment (adamsitnik) [reply src/native/libs/System.Native/pal_io.c:1963]: @tmds thank you!

**Review by stephentoub — APPROVED** [review]

**Comment by adamsitnik** [comment]
/ba-g the failures are known, the DeadLetter was not caused by my changes

**Pull request merged**